libraries/hwlib/library/targets/hwlib-arduino-due.hpp

```diff
@@ -861,6 +895,172 @@ class pin_adc : public hwlib::adc {
    }
 };
 
+class usart_bus {
+private:
+//    Pio & tx_port;
+//    Pio & rx_port;
+//    uint32_t tx_mask;
+//    uint32_t rx_mask;
+
+    Usart * hardwareUSART = nullptr;
+    unsigned int baudrate = BMPTK_BAUDRATE;
+    int n;
+    bool USART_controller_initialized = false;
+
+    Queue<uint8_t, 250> input_buffer;
+   
+    inline uint8_t receive_byte(){
+        return hardwareUSART->US_RHR;
+    }
+
+    unsigned int available() {
+        if (!USART_controller_initialized) {
+            return 0;
+        }
+
+        if ((hardwareUSART->US_CSR & 1) != 0) {
+            input_buffer.push(receive_byte());
+        
+        }
+        return input_buffer.count();
+    }
+
+    void send_byte(const uint8_t &b){
+        while (!tx_ready());
+        hardwareUSART->US_THR = b;
+    }
+    inline bool tx_ready(){
+        return (hardwareUSART->US_CSR & 2);
+    }
+
+public:
+    usart_bus(
+//        uint32_t tx_port_number, 
+//        uint32_t rx_port_number, 
+//        uint32_t tx_pin_number, 
+//        uint32_t rx_pin_number,
+        int n 
+    ):
+//        tx_port{port_registers(tx_port_number)},
+//        rx_port{port_registers(rx_port_number)},
+//        tx_mask{0x1U << tx_pin_number },
+//        rx_mask{0x1U << rx_pin_number },
+        n(n)
+    {
+//        tx_port.PIO_OER = tx_mask;
+//        rx_port.PIO_ODR = rx_mask;
+    }
+
+    usart_bus( uart_ports name ):
+        usart_bus{
+//            uart_port_info( name ).tx.port,
+//            uart_port_info( name ).rx.port,
+//            uart_port_info( name ).tx.pin,
+//            uart_port_info( name ).rx.pin,
+            uart_port_info( name ).n
+            }
+    {}
+
+    ~usart_bus(){
+        disable();
+    }
+
+    void begin(unsigned int boudrate = BMPTK_BAUDRATE){
+        if (USART_controller_initialized) {
+            return;
+        }
+        baudrate = boudrate;
+        if( n == 1 ){
+            hardwareUSART = USART0;
+
+            PIOA->PIO_PDR = PIO_PA10;
+			PIOA->PIO_ABSR &= ~PIO_PA10;
+			PIOA->PIO_PDR = PIO_PA11;
+            PIOA->PIO_ABSR &= ~PIO_PA11;
+
+            PMC->PMC_PCER0 = (0x01 << ID_USART0);
+        
+        } else if (n == 2){
+            hardwareUSART = USART1;
+
+            PIOA->PIO_PDR = PIO_PA12;
+			PIOA->PIO_ABSR &= ~PIO_PA12;
+			PIOA->PIO_PDR = PIO_PA13;
+            PIOA->PIO_ABSR &= ~PIO_PA13;
+
+            PMC->PMC_PCER0 = (0x01 << ID_USART1);
+        } else if (n == 3){
+//          It dont work 
+            hardwareUSART = USART3;
+
+			PIOD->PIO_PDR = PIO_PD4;
+			PIOD->PIO_ABSR |= PIO_PD4;
+			PIOD->PIO_PDR = PIO_PD5;
+            PIOD->PIO_ABSR |= PIO_PD5;
+
+            PMC->PMC_PCER0 = (0x01 << ID_USART3); 
+        } else if (n == 0) {
+            // NO!
```

Comment:
It doesn't look like this should happen. You might wanna give a "HWLIB_PANIC_WITH_LOCATION" here.